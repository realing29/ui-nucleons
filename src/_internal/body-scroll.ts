import { disableBodyScroll, enableBodyScroll, BodyScrollOptions } from 'body-scroll-lock';
import { useEffect } from 'react';

export interface WithBodyScrollLock {
  /** Нужно ли выключать прокрутку body. */
  withScrollDisable?: boolean;

  /** Опции отключения прокрутки. */
  scrollDisableOptions?: BodyScrollOptions;
}

/**
 * Хук управления блокировкой прокрутки body.
 * @param ref Реф элемента для которого нужно отключить прокрутку body.
 * @param needDisable Нужно ли отключать.
 * @param options Опции.
 */
export const useBodyScrollLock = (
  ref: React.RefObject<HTMLElement>,
  needDisable: boolean | undefined,
  options?: BodyScrollOptions,
): void => {
  useEffect(() => {
    const element = ref.current;

    if (element && needDisable) {
      disableBodyScroll(element, options);

      return () => enableBodyScroll(element);
    }
  }, [needDisable, options]);
};
