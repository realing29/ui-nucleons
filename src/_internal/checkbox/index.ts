import { InputHTMLAttributes } from 'react';

export interface CheckboxProps extends InputHTMLAttributes<HTMLInputElement> {
  /** Идентификатор для систем автоматизированного тестирования. */
  'data-testid'?: string;
}
