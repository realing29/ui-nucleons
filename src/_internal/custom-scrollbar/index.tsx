import React from 'react';
import { OverlayScrollbarsComponent, OverlayScrollbarsComponentProps } from 'overlayscrollbars-react';
import classnames from 'classnames/bind';
import styles from './custom-scrollbar.scss';
import 'overlayscrollbars/css/OverlayScrollbars.css';

type Options = NonNullable<OverlayScrollbarsComponentProps['options']>;

export interface CustomScrollbarProps {

  /** Стили. */
  style?: React.CSSProperties

  /** Класс. */
  className?: string

  /** Содержимое. */
  children?: React.ReactNode

  /** Опции переполнения. */
  overflow?: Options['overflowBehavior']

  /** Необходимо поставить true если нужно использовать внутри flexbox-родителя. */
  inFlexBox?: boolean

  /** Сработает при полной прокрутке. */
  onFullScroll?: () => void

  /** Запас для определения полной прокрутки. */
  fullScrollThreshold?: number
}

const cx = classnames.bind(styles);

/**
 * Компонент блока с кастомными полосами прокрутки по гайдам.
 * @param props Свойства.
 * @return Элемент.
 */
export const CustomScrollbar = ({
  style,
  className,
  children,
  overflow,
  inFlexBox = false,
  onFullScroll,
  fullScrollThreshold,
}: CustomScrollbarProps) => (
  <OverlayScrollbarsComponent
    style={style}
    className={cx('custom-scrollbar', className, { 'os-host-flexbox': inFlexBox })}
    options={{
      overflowBehavior: overflow,
      callbacks: onFullScroll
        ? {
          onScrollStop: HandleFullScroll(onFullScroll, fullScrollThreshold),
        }
        : undefined,
    }}
  >
    {children}
  </OverlayScrollbarsComponent>
);

/**
 * Возвращает обработчик прокрутки, который вызовет callback при полной прокрутке.
 * @param callback Callback.
 * @param threshold Запас.
 * @return Обработчик.
 */
export const HandleFullScroll = (
  callback: () => void,
  threshold = 16
) => (event: UIEvent | undefined) => {
  const el = event?.target;

  el instanceof Element
    && el.scrollTop >= el.scrollHeight - el.clientHeight - threshold
    && callback();
};
