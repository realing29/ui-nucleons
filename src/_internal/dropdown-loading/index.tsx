import React from 'react';
import { Spinner } from '../../spinner';
import styles from './index.scss';

type Props = Omit<React.HTMLProps<HTMLDivElement>, 'className'>;

/**
 * Плашка состояния загрузки для Dropdown.
 * @param props Свойства.
 * @return Элемент.
 */
export const DropdownLoading: React.FC<Props> = props => (
  <div {...props} className={styles.root}>
    <Spinner size='small' />
  </div>
);
