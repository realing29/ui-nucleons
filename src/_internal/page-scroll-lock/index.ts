export type { WithPageScrollLock, PageScrollLockAdapterFactory } from './types';
export { BSL_IGNORE_ATTR } from './adapters/body-scroll-lock';
export { usePageScrollLock } from './hook';
export { PageScrollProvider } from './context';
