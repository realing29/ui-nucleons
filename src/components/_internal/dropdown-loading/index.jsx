import React from 'react';
import Spinner from '../../spinner';
import styles from './index.scss';

/**
 * Плашка состояния загрузки для Dropdown.
 * @param {Object} props Свойства.
 * @return {ReactElement} Компонент.
 */
export const DropdownLoading = props => (
  <div {...props} className={styles.root}>
    <Spinner size='small' />
  </div>
);
