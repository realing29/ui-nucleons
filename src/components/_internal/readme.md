# Внутренние компоненты

Компоненты из этого каталога не должны использоваться за пределами данной библиотеки.
