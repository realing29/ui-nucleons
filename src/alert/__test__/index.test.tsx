import React from 'react';
import { act } from 'react-dom/test-utils';
import { mount } from 'enzyme';
import { Alert } from '..';
import { useBodyScrollLock } from '../../_internal/body-scroll';

jest.mock('../../_internal/body-scroll', () => {
  const original = jest.requireActual('../../_internal/body-scroll');

  return {
    ...original,
    __esModule: true,
    useBodyScrollLock: jest.fn(original.useBodyScrollLock),
  };
});

describe('<Alert />', () => {
  it('should renders without props', () => {
    const wrapper = mount(<Alert />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should handle props', () => {
    const wrapper = mount(
      <Alert
        title='Hello, world!'
        withNavBar
        withDivideNavBar={false}
        navBarProps={{ subtitle: 'Subtitle text' }}
        className='test-class'
        children='Main content'
        footer='Footer content'
        inPortal={false}
      />,
    );

    expect(wrapper).toMatchSnapshot();
  });

  it('should disable/enable body scroll', () => {
    expect(useBodyScrollLock).toBeCalledTimes(0);

    mount(<Alert withScrollDisable />);

    expect(useBodyScrollLock).toBeCalledTimes(1);
  });

  it('should do not use disable/enable body scrolling by default', () => {
    mount(<Alert />);

    expect((useBodyScrollLock as any).mock.calls[0][1]).toBe(false);
  });

  it('should call onClose callback on overlay click', () => {
    const spy = jest.fn();
    const wrapper = mount(<Alert onClose={spy} />);

    expect(spy).toBeCalledTimes(0);
    act(() => {
      wrapper.find('[data-testid="alert:overlay"]').simulate('click');
    });
    expect(spy).toBeCalledTimes(1);
  });
});
