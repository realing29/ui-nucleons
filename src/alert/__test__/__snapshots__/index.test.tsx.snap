// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Alert /> should handle props 1`] = `
<Alert
  className="test-class"
  footer="Footer content"
  inPortal={false}
  navBarProps={
    Object {
      "subtitle": "Subtitle text",
    }
  }
  title="Hello, world!"
  withDivideNavBar={false}
  withNavBar={true}
>
  <AlertInner
    className="test-class"
    footer="Footer content"
    navBarProps={
      Object {
        "subtitle": "Subtitle text",
      }
    }
    title="Hello, world!"
    withDivideNavBar={false}
    withNavBar={true}
  >
    <div
      className="overlay"
      data-testid="alert:overlay"
      onClick={[Function]}
      style={
        Object {
          "zIndex": 100,
        }
      }
    >
      <div
        className="alert test-class"
        data-testid="alert"
      >
        <NavBar
          bottomBordered={false}
          className="header"
          subtitle="Subtitle text"
          title="Hello, world!"
        >
          <div
            className="nav-bar header"
            style={
              Object {
                "height": "56px",
              }
            }
          >
            <div
              className="main"
            >
              <div
                className="title truncate"
              >
                Hello, world!
              </div>
              <div
                className="subtitle truncate"
              >
                Subtitle text
              </div>
            </div>
          </div>
        </NavBar>
        <div
          className="main"
        >
          Main content
        </div>
        <div
          className="inner-border-top"
        >
          Footer content
        </div>
      </div>
    </div>
  </AlertInner>
</Alert>
`;

exports[`<Alert /> should renders without props 1`] = `
<Alert>
  <Portal>
    <Portal
      containerInfo={
        <div>
          <div
            class="overlay"
            data-testid="alert:overlay"
            style="z-index: 100;"
          >
            <div
              class="alert"
              data-testid="alert"
            >
              <div
                class="main"
              />
            </div>
          </div>
        </div>
      }
    >
      <AlertInner>
        <div
          className="overlay"
          data-testid="alert:overlay"
          onClick={[Function]}
          style={
            Object {
              "zIndex": 100,
            }
          }
        >
          <div
            className="alert"
            data-testid="alert"
          >
            <div
              className="main"
            />
          </div>
        </div>
      </AlertInner>
    </Portal>
  </Portal>
</Alert>
`;
